### ISSUE TITLE
Question: How to implement computed field by cpu?

### ISSUE THREAD

**Issue by fastengineer** [hasura/graphql-engine#4298]
### Optimal Flow:
On Hasura-UI
- Collecting a form for each computed field by rest api, which gets external data.
### Use case:
For city images I have a table and file_name field in this table.
```ruby
   create_table "city_images", id: :serial, force: :cascade do |t|
    t.integer "city_id", null: false
    t.string "file_name"
    t.index ["city_id"], name: "index_city_pictures_on_city_id"
  end
```

The file_name field has an example "Istanbul.jpg" value.
And these files storing in several s3 bucket.
I have replica buckets in different locations. And I also have bucket for testing and development environment.
In REST API which mounted to /cities url, the application renders all cities as JSON.
In this process the api returns computed full file name url for each city image. Like this:
`https://orangebucket.s3.amazonaws.com/Istanbul.jpg`

### Current Hasura Options:
In Hasura, how I can generate cpu computed fields from rest api?
- Actions. I thought of using the new beta actions. I understand that, it sends out a request for each calculated value. If a list has 10 elements, does that mean that 10 requests will be made? This leads to a N+1 problem. Any ideas?

Thank you.

**Comment by tirumaraiselvan** [comment]
@fastengineer I am slighly confused as to what you want to achieve?

Do you want to populate the table by stripping the full url from the REST API and just having `Istanbul.jpg` in your column ?

Regarding Actions, I am not sure what you mean by one request per element of a list. [Actions](https://hasura.io/docs/1.0/graphql/manual/actions/index.html) are a way to extend Hasura by defining your own types and fields. Each action is backed by a REST API (aka action handler) and  only one request is made to the action handler per invocation.

**Comment by fastengineer** [comment]
@tirumaraiselvan Yes, it is. I am storing just `Istanbul.jpg` in my database.
The rest api application calculates full url: `https://orangebucket.s3.amazonaws.com/Istanbul.jpg`
I need full url in GraphQL response. How I can do it?

If actions is good solution for this case, another question, while using actions:
- Does Hasura calls full url api for each image url.
- If I have collection that contains 100 image. Will Hasura call rest api 100 times for rendering it?

**Comment by sobrinho** [comment]
@fastengineer create a computed field in Hasura.

It allows you to specify a db function that will give you the database row and whatever you return, will be returned in the GraphQL.

**Comment by sobrinho** [comment]
You can do something like:

```sql
CREATE OR REPLACE FUNCTION blah_image_url(row {{table_name}})
  RETURNS varchar
  LANGUAGE sql
  STABLE
AS $function$
  SELECT 'blah' || row.my_column
  ;
$function$;
```

**Comment by fastengineer** [comment]
@sobrinho I know sql functions. Please, read carefully before comment. As I explained above, This process have to done by the rest api application.

**Comment by sobrinho** [comment]
@fastengineer `I need full url in GraphQL response. How I can do it?` and `The file_name field has an example "Istanbul.jpg" value.`, that's how you do it, you don't need actions for that.

You can use [Hasura Computed Fields](https://hasura.io/docs/1.0/graphql/manual/schema/computed-fields.html) that will call the db function in every query to GraphQL.

**Comment by sobrinho** [comment]
Maybe you are not being clear about what you need? :)

**Comment by sobrinho** [comment]
Although you mentioned something about having a REST API that is not clear to me.

You need to pass your record that is being returned by Hasura to this REST API to get the full value because it can't be computed by a db function?

**Comment by sobrinho** [comment]
And for last, if you control the database that Hasura is connected to, can't you simply store the final URL there? It is static or dynamically computed in every call?

This way you don't have to do fancy stuff in Hasura.

**Comment by fastengineer** [comment]
@sobrinho Yes, it's can't be computed by a db function. All of this fields are dynamically computed by application in every call.

Yes, after image upload I can store full url in database. But this solution has some cons about high availability. Our application is very high scale and hosted in multi cloud. S3 buckets can move frequently. Some buckets have read replicas. Moreover, some pictures are not public. Some can only be viewed once. There are records that have been moved to the trash. And it's actually a simple use case, there is more complicated situations. E.g. below.

We have a very large rest api for payment, we dynamically calculate which payment method the customer will use. Money transfers are being processed with more than ten methods, 50+ banks.

About Hasura's remote schemas. A really nice approach. However, some of our rest based micro services that have been developed and stable for 5 years. All of them works with rest, they aren't compatible with graphql

I agree you. Using database functions is best practice when it’s possible. There are places where we use both function and PL / SQL. Unfortunately sometimes it is not enough.

As a result, the new actions feature of the Hasura is very interesting. I think they can work with rest. Maybe it would be very easy to get data from outside and put it into graphql result

**Comment by sobrinho** [comment]
Alright.

I don't have the big picture of the setup but yes, Hasura Actions are the answer here.

Although, try to think also outside the box as well.

If you are in control of this REST API, you might send the id or whatever it needs to the service and it redirects you to the final URL, like `https://my-ha-file-api/{id}/{file} => https://s3.amazon.blah/blah` (or similar). Here you would return the `{api_endpoint}/{file}` from Hasura.

If you are in control of the client that is going to consume this Hasura, you can make a second request from there and leave Hasura outside of it (and even do it in parallel if you can).

Just keep in mind that actions have additionals roundtrips that you may not want to pay: `client -> hasura -> rest api -> hasura -> client`.

And if you are feeling crazy, you can perform http requests from a db function to avoid this as well: https://github.com/pramsey/pgsql-http

I'm not recommending anything here, just showing the options, you are up to decide what is best for your scenario! ❤️

**Comment by sobrinho** [comment]
@fastengineer can this be closed?

**Comment by fastengineer** [comment]
For image I liked redirect psychology, I will give a try. For strong requirements we consider remote-schema and actions. I thought a little bit about the solution.

I added db connections to picture. In the real world it looks like there are 3 server, 1 client, 4 connection. As result of design it will not work fast. It is always best to do so in the database instead of using a new socket if you considering how to scale to millions. I looked at pgsql-http, it is working on sockets, so we have no time saving here. Every time while we use open-internet connection, clients waits, dns server looks for ip address, there is sum of latencies. It is no real time application.

`````` client -> hasura -> db -> rest api -> db -> hasura -> client ``````
We know it's from remote schemas. I am not sure about that Hasura connects db to resolve schema. I guess it does.
`````` client -> hasura -> db -> remote schema -> db -> hasura -> client ``````

If all applications lives in same host server, it'll be fast on unix sockets. But who puts the database and the application on the same server? Private network in datacenter? It is hard to manage.

The best solution when we think a little, we can host an our own Open-Faas container in the same host that Hasura container lives. Then they can communicate very quickly with unix sockets between them.
But database connection latency will still affect us. This is acceptable for database security.
The only backwards this architect is that It might be painful to make a different configuration for each container. Maybe HashiCorp help us.

I think all of these today's casual problems about graphql. Optimal solutions will ramp in next years.
Everybody moves to graphql, because no one don’t want to spend more time on tons of api maintenance anymore. It is incredible. 

I think the problems I have experienced will be experienced in all real applications. If I can get over it somehow early, I would like to share how I did it neatly on a blog. I think a discourse forum will be good fit while learning.
